### PR TITLE
create first draft of branching and release process

### PR DESCRIPTION
**Pull request by KPostOffice**

### CODE REVIEW

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (accorvin) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: What does "deployed to production" mean in our context?

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (accorvin) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: I'm a little bit confused on the use of release branches and tags. This line states that we would create release tags off of main, but that seems to conflict with the existence of release branches. Would we not want to tag off of release branches? If yes, would we always tag at the point of a release? Can a given release branch have multiple tags (i.e. if we need to do multiple builds/releases for a given version

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (accorvin) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: I vote for SemVer

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (accorvin) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md:142]: I think at a minimum we should include a changelog. This can go along with the bullet point below - if there are breaking changes we can explicitly list them in the changelog.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md]: Deployed to production probably isn't accurate in our case, probably just pushed to the `latest` update channels for the operator and the latest tag for controller images.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md]: Release branches would only exist to pull in bug fixes to already released versions. At the time of release you would just create a tag like `v0.1.0`, then if there are some bug fixes that we want to patch onto the minor version we create a branch `release-v0.1` from the `v0.1.0` tag. Then cherry pick commits that we want to include in `v0.1.1`. Once all the commits have been pulled in then we can create a release tag `v0.1.1` off of the `release-v0.1` branch. The idea is that feature and bug dev can happen at the same time without having to pull new features into a patch release.

Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md]: I'll work on the wording

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md:142]: We should also probably at the very least zip up the source code so that the code can be downloaded without using `git`

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md]: I agree. I think the only other sane option is calendar versioning, but that doesn't really jive with the idea of patch releases and doesn't capture potentially breaking changes the way semver does.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: I think we could clarify this a little, the prow plugin we are using will require us to have an `approved` label and `lgtm` label on PRs. In general the preference is for those two labels to come from different people.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]:
```suggestion
- When the release branch is ready for deployment, create a tag on GitHub with the patch version incremented. For example, `v1.2.0` will become v1.2.1`. This will trigger the deployment pipeline for that tag.
```

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]:
```suggestion
```

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: Just to further highlight our usage of semver, I'd change references of 1.3 to 1.3.0, etc.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md:145]: Is there a CodeFlare mail list? At a minimum, we can share these out to the slack channel

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]:
```suggestion
more involved merging. One benefit that it does offer is a slower dev process for junior devs.
```

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]:
```suggestion
| CodeFlare SDK                 | Mustafa Eyceoz | 4/19/2023  | yes       |
| MCAD                          | Abhishek Malvankar | 4/19/2023  | yes       |
| InstaScale                    | Abhishek Malvankar | 4/19/2023  | yes       |
| CodeFlare Operator            | Anish Asthana | 4/19/2023  | yes       |
```

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]: We should add at least Alex, myself, Mustafa, and Abhishek to this list of reviewers. Additional people are welcome to review.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (accorvin) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md]: This makes a lot of sense. Thanks for the clarification!

**Comment by accorvin** [comment]
I'm good with this pending the existing comments.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (Maxusmusti) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md:145]: there is a projectcodeflare@gmail.com account unused currently, also Carlos has the codeflare.dev domain so a new email could be created in that (if we decide that Slack isn't enough as-is atm)

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (tedhtchang) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md]:
```suggestion
- Setting a release cycle for the Codeflare project.
```

**Review by jbusche — DISMISSED** [review]
Thanks Kevin, this is a great start, and LGTM

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (tedhtchang) [inline PCF-ADR-0002-release-story-and-branch-maintenance.md:81]: We may want to store the artifacts on a stagging area for a few days for people to download and validate as a last gate before going out.

Comment (KPostOffice) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md:81]: I think this is definitely doable. One could create `v1.1.0-alpha` tag and then pull any bug fixes that are made into the `v1.1` branch as normal. The second tag on that branch would be `v1.1.0` or even `v1.0.0-beta`. I'll add this to the dev process as an option for staging releases.

**Comment by jbusche** [comment]
Shoot - it occurs to me that we should have vulnerability scans as part of the release process.  It'd be great if a report could be generated for each build, but I don't think we should have it fail if/when it finds things because that could really hold up people's PRs and releases, especially if a last-minute vulnerability came out.

**Comment by KPostOffice** [comment]
> Shoot - it occurs to me that we should have vulnerability scans as part of the release process. It'd be great if a report could be generated for each build, but I don't think we should have it fail if/when it finds things because that could really hold up people's PRs and releases, especially if a last-minute vulnerability came out.

I'm going to leave specifics such as image/artifact building and scanning out of this ADR, both because I'm not familiar with all the project specifics and because I wanted to keep this ADR focused. I'll add it to the open questions section.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (tedhtchang) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md:81]: Looks good. Well explained.

File: PCF-ADR-0002-release-story-and-branch-maintenance.md
Comment (anishasthana) [reply PCF-ADR-0002-release-story-and-branch-maintenance.md:145]: I'll start a thread in slack too.

**Review by anishasthana — APPROVED** [review]
/lgtm

**Pull request merged**
